src/java.base/share/classes/java/lang/StringConcatHelper.java

```diff
@@ -39,6 +41,91 @@
  * combinators there.
  */
 final class StringConcatHelper {
+    static class StringConcatBase {
+        @Stable final String[] constants;
+        final int length;
+        final byte coder;
+        protected StringConcatBase(String[] constants) {
+            int length = 0;
+            byte coder = String.LATIN1;
+            this.constants = constants;
+            for (String c : constants) {
+                length += c.length();
+                coder |= c.coder();
+            }
+            this.length = length;
+            this.coder = coder;
+        }
+
+        @ForceInline
+        private String concat0(String value) {
```

Comment:
I think we should make the base class abstract and add a `final class StringConcat1` with all these single-argument concat private methods, and that class is exposed via JLA instead.